x/leverage/keeper/keeper.go

```diff
@@ -271,3 +265,181 @@ func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
 	key := types.CreateCollateralSettingKey(borrowerAddr, denom)
 	return store.Has(key)
 }
+
+// LiquidateBorrow attempts to repay one of an eligible borrower's borrows (in part or in full) in exchange
+// for a selected denomination of uToken collateral. If the borrower is not over their borrow limit, or
+// the repayment or reward denominations are invalid, an error is returned. If the attempted repayment
+// is greater than the amount owed or the maximum that can be repaid due to parameters (close factor)
+// then a partial liquidation, equal to the maximum valid amount, is performed. The same occurs if the
+// value of collateral in the selected reward denomination cannot cover the proposed repayment.
+// Because partial liquidation is possible and exchange rates vary, LiquidateBorrow returns the actual
+// amount of tokens repaid and uTokens rewarded (in that order).
+func (k Keeper) LiquidateBorrow(
+	ctx sdk.Context, liquidatorAddr, borrowerAddr sdk.AccAddress, repayment sdk.Coin, rewardDenom string,
+) (sdk.Int, sdk.Int, error) {
+	if !repayment.IsValid() {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrInvalidAsset, repayment.String())
+	}
+	if !k.IsAcceptedUToken(ctx, rewardDenom) {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrInvalidAsset, rewardDenom)
+	}
+
+	// Get total borrowed by borrower (all denoms)
+	borrowed, err := k.GetBorrowerBorrows(ctx, borrowerAddr)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Get borrower uToken balances, for all uToken denoms enabled as collateral
+	collateral := k.GetBorrowerCollateral(ctx, borrowerAddr)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Use oracle helper functions to find total borrowed value in USD
+	borrowValue, err := k.TotalPrice(ctx, borrowed)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Use collateral weights to compute borrow limit from enabled collateral
+	borrowLimit, err := k.CalculateBorrowLimit(ctx, collateral)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Confirm borrower's eligibility for liquidation
+	if borrowLimit.GTE(borrowValue) {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrLiquidationIneligible, borrowerAddr.String())
+	}
+
+	// Get reward-specific incentive and dynamic close factor
+	baseRewardDenom := k.FromUTokenToTokenDenom(ctx, rewardDenom)
+	liquidationIncentive, closeFactor, err := k.LiquidationParams(ctx, baseRewardDenom, borrowValue, borrowLimit)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Get liquidator's available balance of base asset to repay
+	liquidatorBalance := k.bankKeeper.GetBalance(ctx, liquidatorAddr, repayment.Denom)
+	// Return specific error on zero liquidator balance
+	if liquidatorBalance.Amount.IsZero() {
```

Comment:
note, the x/bank module will never have a balance of zero for an address. So this check doesn't hurt, but it's not necessary.